tests/pytest_tests/unit_tests/test_artifacts/test_storage.py

```diff
@@ -36,6 +39,61 @@ def test_opener_rejects_append_mode(artifact_file_cache):
         f.write("example")
 
 
+def test_opener_works_across_filesystem_boundaries(
+    tmp_path, artifact_file_cache, fs: FakeFilesystem
+):
+    # This is not ideal, as we'd much rather test e.g. `Artifact.download()` directly or E2E.
+    #
+    # However, we're using `pyfakefs` to mock mounted/partitioned filesystems, and it doesn't play well with
+    # some of the internals of ArtifactFileCache without additional, potentially brittle/complicated patches
+    # to unrelated operations (e.g. around `subprocess.call()`).  This will have to do for the moment.
+
+    # Some setup/patching we have to do to get this test to play well with `pyfakefs`
+    # Note: Cast to str looks redundant, but is intentional (for python < 3.10).
+    # https://pytest-pyfakefs.readthedocs.io/en/latest/troubleshooting.html#pathlib-path-objects-created-outside-of-tests
+    fake_tmp_path = Path(str(tmp_path))
+    fs.create_dir(fake_tmp_path)
+
+    fake_cache_dir = Path(str(artifact_file_cache._cache_dir))
+    fake_cache_obj_dir = Path(str(artifact_file_cache._obj_dir))
+    fake_cache_temp_dir = Path(str(artifact_file_cache._temp_dir))
+    artifact_file_cache._cache_dir = fake_cache_dir
+    artifact_file_cache._obj_dir = fake_cache_obj_dir
+    artifact_file_cache._temp_dir = fake_cache_temp_dir
+    fs.create_dir(fake_cache_dir)
+    fs.create_dir(fake_cache_obj_dir)
+    fs.create_dir(fake_cache_temp_dir)
+
+    cache_path, _, cache_opener = artifact_file_cache.check_md5_obj_path(
+        example_digest, 7
+    )
+    with cache_opener() as f:
+        f.write("test-123")
+
+    # Simulate a destination filepath on the mounted filesystem
+    dest_dir = fake_tmp_path / "mount"
+    dest_path = dest_dir / "dest.txt"
+    fs.create_dir(dest_dir)
+    fs.add_mount_point(str(dest_dir))
+
+    # Sanity check: `os.rename` should fail across the (fake) filesystem boundary
+    # This is extra assurance that we're still testing what we expect to test
+    with pytest.raises(OSError) as excinfo:
+        os.rename(cache_path, dest_path)
+        assert excinfo.value.args[0] == errno.EXDEV
```

Comment:
the `assert` does not run if `os.rename` raises an exception

```suggestion
    with pytest.raises(OSError) as excinfo:
        os.rename(cache_path, dest_path)
    assert excinfo.value.args[0] == errno.EXDEV
```

Reply:
Whoops, you're right.  thanks for the catch!